xocto/ranges.py

```diff
@@ -945,3 +946,56 @@ def any_overlapping(ranges: Iterable[Range[T]]) -> bool:
             return True
         range_set.add(range)
     return False
+
+
+def as_finite_datetime_periods(
+    periods: Iterable[HalfFiniteDatetimeRange | DatetimeRange],
+) -> Sequence[FiniteDatetimeRange]:
+    """
+    Casts the given date/time periods as finite periods.
+
+    This is useful when working with potentially infinite ranges that are
+    known to be finite e.g. due to intersection with a finite range.
+
+    Raises:
+        ValueError: If one or more periods is not finite.
+    """
+    finite_periods = []
+
+    for period in periods:
+        if period.start is None or period.end is None:
+            raise ValueError("Period is not finite at start or end or both")
+
+        finite_periods += [FiniteDatetimeRange(period.start, period.end)]
+
+    return finite_periods
+
+
+def as_date_periods(
+    periods: Iterable[FiniteDatetimeRange],
+    *,
+    date_converter: Callable[[datetime.datetime], datetime.date],
+) -> Sequence[FiniteDateRange]:
+    """
+    Converts the given date/time periods to date periods.
+
+    The given date converter is used to decide what date to ascribe to a
+    given date/time, since we lose timezone info in this conversion.
```

Comment:
🤔 I don't quite understand this part, could you clarify the following:

My understanding is that we should be passing around datetimes which have timezone information everywhere, in which case we can use `localtime.date` to get the correct day as needed.

I guess the issue is that the typing doesn't prevent us from passing in an unzoned datetime, but maybe we could just throw a value error in that case?